mesonbuild/compilers.py

```diff
@@ -2729,3 +2735,137 @@ def unix_compile_flags_to_native(self, args):
+        std = options['c_std']
+        if std.value != 'none':
+            args.append('-std=' + std.value)
+        return args
+
+    def get_std_shared_lib_link_args(self):
+        return ['-shared']
+
+    def has_multi_arguments(self, args, env):
+        return super(IntelCCompiler, self).has_multi_arguments(args + ['-diag-error', '10006'], env)
+
+
+class IntelCPPCompiler(IntelCompiler, CPPCompiler):
+
+    def __init__(self, exelist, version, icc_type, is_cross, exe_wrap):
+        CPPCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+        IntelCompiler.__init__(self, icc_type)
+        self.warn_args = {'1': ['-Wall', '-w3', '-diag-disable:remark', '-Wpch-messages', '-Wnon-virtual-dtor'],
+                          '2': ['-Wall', '-w3', '-diag-disable:remark', '-Wextra', '-Wpch-messages', '-Wnon-virtual-dtor'],
+                          '3': ['-Wall', '-w3', '-diag-disable:remark', '-Wpedantic', '-Wextra', '-Wpch-messages', '-Wnon-virtual-dtor']}
+
+    def get_options(self):
+        if mesonlib.version_compare(self.version, '>=17.0.0'):
+            opts = {'cpp_std': coredata.UserComboOption('cpp_std', 'C++ language standard to use',
+                                                        ['none', 'c++11', 'c++0x', 'c++14',
+                                                         'gnu++98', 'gnu++11', 'gnu++0x', 'gnu++14'],
+                                                        'none'),
+                    'cpp_debugstl': coredata.UserBooleanOption('cpp_debugstl',
+                                                               'STL debug mode',
+                                                               False)}
+        else:
+            opts = {'cpp_std': coredata.UserComboOption('cpp_std', 'C++ language standard to use',
+                                                        ['none', 'c++03', 'c++11', 'c++0x',
+                                                         'gnu++98', 'gnu++03', 'gnu++11', 'gnu++0x'],
+                                                        'none'),
+                    'cpp_debugstl': coredata.UserBooleanOption('cpp_debugstl',
+                                                               'STL debug mode',
+                                                               False)}
+        return opts
+
+    def get_option_compile_args(self, options):
+        args = []
+        std = options['cpp_std']
+        if std.value != 'none':
+            args.append('-std=' + std.value)
+        if options['cpp_debugstl'].value:
+            args.append('-D_GLIBCXX_DEBUG=1')
+        return args
+
+    def get_option_link_args(self, options):
+        return []
+
+    def get_compiler_check_args(self):
+        # -fpermissive allows non-conforming code to compile which is necessary
+        # for many C++ checks. Particularly, the has_header_symbol check is
+        # too strict without this and always fails.
+        return self.get_no_optimization_args() + ['-fpermissive']
```

Comment:
I get the same test results with and without `-fpermissive`.
You tested it with ICC15 right? So should we just leave it out?